### PR TITLE
Fix fatal error on HHVM with nullable type

### LINKED ISSUE #203: Error when running under HHVM
When running a test that uses Phony with HHVM, this error appears:

```
Fatal error: Syntax only allowed in Hack files (<?hh) or with -v Eval.EnableHipHopSyntax=true in vendor/eloquent/phony/src/Reflection/FeatureDetector.php(533) : eval()'d code on line 1
```

Job source https://travis-ci.org/thephpleague/oauth2-client/jobs/173009708

I find this particularly odd, since I have used Phony in another project that tests on HHVM without issues. Any idea what the cause is?

### PR DESCRIPTION
**Pull request by shadowhand**
HHVM does not support the nullable type outside of Hack code.

Broken by 6bf897cb2215ff209ba5e2794c55faaa63d6b4b4

Fixes #203
Refs thephpleague/oauth2-client#584

### CODE REVIEW

**Comment by ezzatron** [comment]
Thanks for the PR! Looks good, and should be fine to merge.

I'd just like to check a couple of things first, and add some extra tests if possible. Is there any chance you can point me at the actual code being mocked? Does it just use nullable types in a parameter?

**Comment by ezzatron** [comment]
So I've been doing a bit of poking around, and it turns out that nullable types are tested and apparently working against multiple versions of HHVM already, so there must be something else funky going on here.

I noticed that your linked repo's Travis CI config is using a version of HHVM that (I think) Travis are moving away from, and may not have been updated in some time. Phony uses something like this to test multiple versions of HHVM ([related Travis CI documentation](https://docs.travis-ci.com/user/languages/php/#HHVM-versions)):

```
matrix:
    include:
        - php: 5.3
        - php: 5.4
        - php: 5.5
        - php: 5.6
        - php: 7.0
        - php: 7.1
        - php: nightly
        - php: hhvm-3.6
          sudo: required
          dist: trusty
          group: edge
        - php: hhvm-3.9
          sudo: required
          dist: trusty
          group: edge
        - php: hhvm-3.12
          sudo: required
          dist: trusty
          group: edge
        - php: hhvm-3.15
          sudo: required
          dist: trusty
          group: edge
        - php: hhvm-nightly
          sudo: required
          dist: trusty
          group: edge
    fast_finish: true
    allow_failures:
        - php: 7.1
        - php: nightly
        - php: hhvm-nightly
```

I'm not saying the problem you found isn't a bug, but I'm guessing it only affects some older versions of HHVM 3.6. Since HHVM will probably eventually support nullable types in a way that is compatible with PHP7, I don't just want to remove the feature for ALL versions of HHVM.

I might open a PR against your `feature/replace-mockery-with-phony` branch with an updated Travis CI config, and see if it still breaks. If so, at least I can track down the root cause.

**Comment by ezzatron** [comment]
So I got some... *interesting* results. I created a fork of `oauth2-client`, updated the Travis config, and reverted to using the latest version of Phony. On [this build](https://travis-ci.org/ezzatron/oauth2-client/builds/173144588), it seems like it doesn't work for 3.6, does work for 3.9 and 3.12, and then doesn't work again for 3.15. Pretty strange:

screen shot 2016-11-04 at 12 40 13 pm

I also don't understand how this code is working fine in Phony's own builds, and fataling elsewhere.

**Comment by shadowhand** [comment]
I was able to narrow down the failure to this mock. At least it is the first one that fails. Other mocks that use similar code also cause the same failure locally:

```
$stream = Phony::mock(StreamInterface::class);
$stream->__toString->returns(
    '{"access_token": "mock_access_token", "expires": 3600, "refresh_token": "mock_refresh_token", "uid": 1}'
);
```

Here is the output of `hhvm --version`:

```
HipHop VM 3.15.2 (rel)
Compiler: 1478195283_N
Repo schema: 8c26071f79c5ac5704473dbb4dc2ff4b22c10280
```

**Comment by ezzatron** [comment]
@shadowhand Thanks!

I tried extracting the code that's causing the fatal and running on 3v4l.org: https://3v4l.org/d0Yhm

TL;DR It seems like HHVM's behavior is pretty unstable here, and there appears to be no version of HHVM that truly supports PHP7-compatible nullable types. There are just some versions where the syntax doesn't cause a fatal.

Given that, I'm inclined to just merge this PR as-is. It's a bit of a shame, because if HHVM does end up adding support in later versions, Phony will have to use a version check rather than real feature detection. I try to avoid that where possible.

Anyway, does that sound reasonable to you?

**Comment by shadowhand** [comment]
Definitely seems reasonable, considering that the failure crosses multiple HHVM versions. Really strange that it fails on older versions *and* newer versions, but worked for a couple of versions.

Seems like this should be reported upstream to HHVM. Will you take care of that, since you have a reproducible test case?

**Comment by shadowhand** [comment]
PS: I dunno what's going on with Appveyor here, it has been stalled since I created the PR.

**Comment by ezzatron** [comment]
Yeah, I should probably do that. It would be nice to know what their plans are regarding PHP7 compatibility anyway. I'll link back here so you can follow.

Don't worry about AppVeyor, I'll sort it out.

**Comment by shadowhand** [comment]
Cheers, man. If we ever happen to meet up, I owe you a beer. 🍻

**Comment by ezzatron** [comment]
@shadowhand No worries, I also owe you one! It's great to see Phony being used in the wild on some real projects. If you're ever in Australia! 🇦🇺 

FYI HHVM issue is facebook/hhvm#7467

**Comment by ezzatron** [comment]
@shadowhand Available in [0.14.1](https://github.com/eloquent/phony/releases/0.14.1)
